File: pageindicatorview/src/main/java/com/rd/draw/controller/MeasureController.java
package com.rd.draw.controller;

import androidx.annotation.NonNull;
import android.util.Pair;
import android.view.View;
import com.rd.animation.type.AnimationType;
import com.rd.draw.data.Indicator;
import com.rd.draw.data.Orientation;

public class MeasureController {

    public Pair<Integer, Integer> measureViewSize(@NonNull Indicator indicator, int widthMeasureSpec, int heightMeasureSpec) {
        int widthMode = View.MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = View.MeasureSpec.getSize(widthMeasureSpec);

        int heightMode = View.MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = View.MeasureSpec.getSize(heightMeasureSpec);

        int count = indicator.getCount();
        int radius = indicator.getRadius();
        int stroke = indicator.getStroke();

        int padding = indicator.getPadding();
        int paddingLeft = indicator.getPaddingLeft();
        int paddingTop = indicator.getPaddingTop();
        int paddingRight = indicator.getPaddingRight();
        int paddingBottom = indicator.getPaddingBottom();

        int circleDiameterPx = radius * 2;
        int desiredWidth = 0;
        int desiredHeight = 0;

        int width;
        int height;

        Orientation orientation = indicator.getOrientation();
        if (count != 0) {
            int diameterSum = circleDiameterPx * count;
            int strokeSum = (stroke * 2) * count;

            int paddingSum = padding * (count - 1);
            int w = diameterSum + strokeSum + paddingSum;
            int h = circleDiameterPx + stroke;

            if (orientation == Orientation.HORIZONTAL) {
                desiredWidth = w;
                desiredHeight = h;

            } else {
                desiredWidth = h;
                desiredHeight = w;
            }
        }

        if (indicator.getAnimationType() == AnimationType.DROP) {
            if (orientation == Orientation.HORIZONTAL) {
                desiredHeight *= 2;
            } else {
                desiredWidth *= 2;
            }
        }

        int horizontalPadding = paddingLeft + paddingRight;
        int verticalPadding = paddingTop + paddingBottom;

        if (orientation == Orientation.HORIZONTAL) {
            desiredWidth += horizontalPadding;
            desiredHeight += verticalPadding;

        } else {
            desiredWidth += horizontalPadding;
            desiredHeight += verticalPadding;
        }

        if (widthMode == View.MeasureSpec.EXACTLY) {
            width = widthSize;
        } else if (widthMode == View.MeasureSpec.AT_MOST) {
            width = Math.min(desiredWidth, widthSize);
        } else {
            width = desiredWidth;
        }

        if (heightMode == View.MeasureSpec.EXACTLY) {
            height = heightSize;
        } else if (heightMode == View.MeasureSpec.AT_MOST) {
            height = Math.min(desiredHeight, heightSize);
        } else {
            height = desiredHeight;
        }

        if (width < 0) {
            width = 0;
        }

        if (height < 0) {
            height = 0;
        }

        indicator.setWidth(width);
        indicator.setHeight(height);

        return new Pair<>(width, height);
    }
}

File: pageindicatorview/src/main/java/com/rd/draw/controller/MeasureController.kt
package com.rd.draw.controller

import android.util.Pair
import android.view.View
import com.rd.animation.type.AnimationType
import com.rd.draw.data.Indicator
import com.rd.draw.data.Orientation

class MeasureController {

	fun measureViewSize(indicator: Indicator, widthMeasureSpec: Int, heightMeasureSpec: Int): Pair<Int, Int> {
		val widthMode = View.MeasureSpec.getMode(widthMeasureSpec)
		val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)

		val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)
		val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)

		val count = indicator.count
		val radius = indicator.radius
		val stroke = indicator.stroke

		val padding = indicator.padding
		val paddingLeft = indicator.paddingLeft
		val paddingTop = indicator.paddingTop
		val paddingRight = indicator.paddingRight
		val paddingBottom = indicator.paddingBottom

		val circleDiameterPx = radius * 2
		var desiredWidth = 0
		var desiredHeight = 0

		var width = 0
		var height = 0

		val orientation = indicator.orientation
		if (count != 0) {
			val diameterSum = circleDiameterPx * count
			val strokeSum = stroke * 2 * count

			val paddingSum = padding * (count - 1)
			val w = diameterSum + strokeSum + paddingSum
			val h = circleDiameterPx + stroke

			if (orientation === Orientation.HORIZONTAL) {
				desiredWidth = w
				desiredHeight = h

			} else {
				desiredWidth = h
				desiredHeight = w
			}
		}

		if (indicator.animationType == AnimationType.DROP) {
			if (orientation === Orientation.HORIZONTAL) {
				desiredHeight *= 2
			} else {
				desiredWidth *= 2
			}
		}

		val horizontalPadding = paddingLeft + paddingRight
		val verticalPadding = paddingTop + paddingBottom

		if (orientation === Orientation.HORIZONTAL) {
			desiredWidth += horizontalPadding
			desiredHeight += verticalPadding

		} else {
			desiredWidth += horizontalPadding
			desiredHeight += verticalPadding
		}

		when (widthMode) {
			View.MeasureSpec.EXACTLY -> width = widthSize
			View.MeasureSpec.AT_MOST -> width = Math.min(desiredWidth, widthSize)
			View.MeasureSpec.UNSPECIFIED -> width = desiredWidth
		}

		when (heightMode) {
			View.MeasureSpec.EXACTLY -> height = heightSize
			View.MeasureSpec.AT_MOST -> height = Math.min(desiredHeight, heightSize)
			View.MeasureSpec.UNSPECIFIED -> height = desiredHeight
		}

		if (width < 0) width = 0
		if (height < 0) height = 0

		indicator.width = width
		indicator.height = height

		return Pair(width, height)
	}
}
